### PR TITLE
feat(probe): implement mount probe

### PR DESCRIPTION
**Pull request by akhilerm**
Implement mount probe to fetch mount related information like filesystem, mount point etc.

These are the BlockDevice CR from various setups after implementing mountprobe

**1.** Local cluster in minikube, with external storage just attached
```
apiVersion: openebs.io/v1alpha1
kind: BlockDevice
metadata:
  creationTimestamp: "2019-05-10T10:03:51Z"
  generation: 1
  labels:
    kubernetes.io/hostname: minikube
    ndm.io/blockdevice-type: blockdevice
  name: blockdevice-adb1b23f7ba395988d029d78ef7bda58
  resourceVersion: "22533"
  selfLink: /apis/openebs.io/v1alpha1/blockdevices/blockdevice-adb1b23f7ba395988d029d78ef7bda58
  uid: e92d8f2f-730a-11e9-aa6a-9840bb3de22d
spec:
  capacity:
    logicalSectorSize: 512
    physicalSectorSize: 0
    storage: 15502147584
  details:
    compliance: ""
    deviceType: ""
    firmwareRevision: ""
    model: DataTraveler_3.0
    serial: 50E5495131BBB060892FBC8E
    vendor: Kingston
  devlinks:
  - kind: by-id
    links:
    - /dev/disk/by-id/usb-Kingston_DataTraveler_3.0_50E5495131BBB060892FBC8E-0:0
  - kind: by-path
    links:
    - /dev/disk/by-path/pci-0000:00:14.0-usb-0:3:1.0-scsi-0:0:0:0
  filesystem: {}
  partitioned: "No"
  path: /dev/sdc
status:
  claimState: Unclaimed
  state: Active
```

**2.** Unformatted/Unmounted GPD attached to a node
```
apiVersion: openebs.io/v1alpha1
kind: BlockDevice
metadata:
  creationTimestamp: "2019-05-10T10:12:07Z"
  generation: 1
  labels:
    kubernetes.io/hostname: gke-akhil-local-pv-default-pool-7dd3a726-g2q3
    ndm.io/blockdevice-type: blockdevice
  name: blockdevice-71b104c0ee3d3892331e7eeb8c7860e6
  resourceVersion: "215111"
  selfLink: /apis/openebs.io/v1alpha1/blockdevices/blockdevice-71b104c0ee3d3892331e7eeb8c7860e6
  uid: 1138b70f-730c-11e9-856d-42010a8001a0
spec:
  capacity:
    logicalSectorSize: 512
    physicalSectorSize: 0
    storage: 107374182400
  details:
    compliance: ""
    deviceType: ""
    firmwareRevision: ""
    model: PersistentDisk
    serial: persistent-disk-2
    vendor: Google
  devlinks:
  - kind: by-id
    links:
    - /dev/disk/by-id/scsi-0Google_PersistentDisk_persistent-disk-2
    - /dev/disk/by-id/google-persistent-disk-2
  - kind: by-path
    links:
    - /dev/disk/by-path/pci-0000:00:03.0-scsi-0:0:2:0
  filesystem: {}
  partitioned: "No"
  path: /dev/sdc
status:
  claimState: Unclaimed
  state: Active
```

**3.** local SSD in GKE
```
apiVersion: openebs.io/v1alpha1
kind: BlockDevice
metadata:
  creationTimestamp: "2019-05-10T10:09:45Z"
  generation: 1
  labels:
    kubernetes.io/hostname: gke-akhil-local-pv-default-pool-7dd3a726-g2q3
    ndm.io/blockdevice-type: blockdevice
  name: blockdevice-8f0ae3211abab14084bee151f6bcabe5
  resourceVersion: "214794"
  selfLink: /apis/openebs.io/v1alpha1/blockdevices/blockdevice-8f0ae3211abab14084bee151f6bcabe5
  uid: bca2da84-730b-11e9-856d-42010a8001a0
spec:
  capacity:
    logicalSectorSize: 4096
    physicalSectorSize: 4096
    storage: 402653184000
  details:
    compliance: SPC-4
    deviceType: SSD
    firmwareRevision: ""
    model: EphemeralDisk
    serial: local-ssd-0
    vendor: Google
  devlinks:
  - kind: by-id
    links:
    - /dev/disk/by-id/scsi-0Google_EphemeralDisk_local-ssd-0
    - /dev/disk/by-id/google-local-ssd-0
  - kind: by-path
    links:
    - /dev/disk/by-path/pci-0000:00:04.0-scsi-0:0:1:0
  filesystem:
    fsType: ext4
    mountPoint: /mnt/disks/ssd0
  partitioned: "No"
  path: /dev/sdb
status:
  claimState: Unclaimed
  state: Active
```

### CODE REVIEW

File: pkg/udev/mockdata.go [hunk 1]
```diff
@@ -83,6 +84,7 @@ func MockDiskDetails() (MockOsDiskDetails, error) {
 	diskDetails.Wwn = device.GetPropertyValue(UDEV_WWN)
 	diskDetails.Uid = device.GetUid()
 	diskDetails.FileSystem = UDEV_FS_NONE
+	diskDetails.Mountpoint = "/" // always take the disk mounted at /
```
Comment (kmova) [inline pkg/udev/mockdata.go]: Let the default option be empty ( `""`), as the default FileSystem is none.

Comment (akhilerm) [reply pkg/udev/mockdata.go]: @kmova  Should I change it everywhere, we use Filesystem type? If no filesystem exists, the CR will have not have the field., But internally, in the `DiskInfo` we pass it as None.
Also, should the default mountpoint also be removed?

File: cmd/ndm_daemonset/probe/mountprobe.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
+package probe
```
Comment (imazik) [inline cmd/ndm_daemonset/probe/mountprobe.go]: licence header missing here.
@kmova can we have same licence check which is in `openebs/maya`.

File: pkg/mount/mountutil.go [hunk 3]
```diff
@@ -110,3 +116,28 @@ func getDiskDevPath(partition string) (string, error) {
 	}
 	return "/dev/" + diskName, nil
 }
+
+// getPartitionName gets the partition name from the mountpoint. Each line of a mounts file
+// is passed to the function, which is parsed and partition name is obtained
+func (m *MountUtil) getPartitionName(mountLine string) (DeviceMountAttr, bool) {
+	mountAttr := DeviceMountAttr{}
+	isValid := false
+	if parts := strings.Split(mountLine, " "); parts[1] == m.mountPoint {
```
Comment (imazik) [inline pkg/mount/mountutil.go]: nit: can we have length check to avoid npe

File: pkg/udev/mockdata.go (hunk 1 above)
Comment (kmova) [reply pkg/udev/mockdata.go]: If we have it set as "None", can we keep it as empty.. and then data structure set as omit if empty?

Comment (akhilerm) [reply pkg/udev/mockdata.go]: @kmova  I have already set it as `omitempty`. But omitempty works only if the struct is a pointer right?. "None" is used only inside the code for easier comparisons. The value in CR will be empty.
But the struct will show it as `FileSystemInfo:{}`

File: cmd/ndm_daemonset/probe/mountprobe.go (hunk 2 above)
Comment (akhilerm) [reply cmd/ndm_daemonset/probe/mountprobe.go]: Added license header to new files

File: pkg/mount/mountutil.go (hunk 3 above)
Comment (akhilerm) [reply pkg/mount/mountutil.go]: Thanks @imazik . Fixed.

File: pkg/udev/mockdata.go (hunk 1 above)
Comment (kmova) [reply pkg/udev/mockdata.go]: cool. got it. This function is about OS disk and assumption is made that it is mounted under `/`. Only used for unit tests.

File: pkg/mount/mountutil.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	if parts := strings.Split(mountLine, " "); parts[1] == m.mountPoint {
+		mountAttr.DevPath = strings.Replace(parts[0], "/dev/", "", 1)
+		isValid = true
+	}
+	return mountAttr, isValid
+}
+
+// getMountName gets the mountpoint, filesystem etc from the partition name. Each line of a mounts
+// file is passed to the function, which is parsed to get the information
```
Comment (kmova) [inline pkg/mount/mountutil.go]: Provide an example of the mountline. Will there be a cases, where devices can be mounted with mount options and are those handled by this function?

Comment (akhilerm) [reply pkg/mount/mountutil.go]: @kmova  when devices are mounted with mount options, those options will be ignored currently.

Comment (akhilerm) [reply pkg/mount/mountutil.go]: Provided sample `mountline`

**Pull request merged**